### ISSUE TITLE
Can add `R.flatten` the ability of flattening object?

### ISSUE THREAD

**Issue by adispring** [ramda/ramda#1984]

**Comment by CrossEye** [comment]
Can you describe what you would want?  Specifically, how would it relate to the existing `flatten` on nested lists?  (For if they are not closely related, we would not use the same function for them.)  How would you handle objects that contain lists?

**Comment by kurtmilam** [comment]
Maybe something along these lines:
https://github.com/ramda/ramda/wiki/Cookbook#flatten-a-nested-object-into-dot-separated-key--value-pairs
?

**Comment by kurtmilam** [comment]
Maybe something along [these lines](https://github.com/ramda/ramda/wiki/Cookbook#flatten-a-nested-object-into-dot-separated-key--value-pairs) (from the wiki / Cookbook)?

```javascript
flattenObj({a:1, b:{c:3}, d:{e:{f:6}, g:[{h:8, i:9}, 0]}})
//=> {"a": 1, "b.c": 3, "d.e.f": 6, "d.g.0.h": 8, "d.g.0.i": 9, "d.g.1": 0}
```

That solution creates one element per leaf node in the input object. Another solution would return an element for every element in the input object, like this:
```javascript
flattenObj({a:1, b:{c:3}, d:{e:{f:6}, g:[{h:8, i:9}, 0]}})
//=> {"a": 1, "b": {}, "b.c": 3, "d": {}, "d.e": {}, "d.e.f": 6, "d.g": [], "d.g.0": {}, "d.g.0.h": 8, "d.g.0.i": 9, "d.g.1": 0}
```
I wrote and use a `flattenObject` function that works like the second example.

**Comment by adispring** [comment]
```js
const data = {a: 1, b: {c: 2, d: {e: 3}}}
flattenObj(data) = {a: 1, c: 2, e: 3} 
```
the flattened result only contains item that not contains child 'nested' object.

**Comment by kurtmilam** [comment]
@adispring What should the function do when keys collide? For instance:
```javascript
const data = {a: 1, b: {a: 2}}
flattenObj(data) = ?
```

**Comment by adispring** [comment]
That's a problem, @kurtmilam, your solution is useful enough for me, and I'll look through the Cookbook.

**Comment by CrossEye** [comment]
@kurtmilam: Could you describe how you use this?  I've never had any need for anything remotely like this.

Would you worry about the clash of keys for an object that has a "foo.bar" key and a "foo" key whose value has a "bar" one?

**Comment by ascartabelli** [comment]
@CrossEye Years ago I wrote and used a function like the first one @kurtmilam mentioned in a [MVP](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) architecture to tie models (nested JSON) to HTML ids (flattened keys) and vice-versa.
Cleaned up the Presenters' code, did the job quickly, but not a clean solution overall; and not one to be exposed in a public library, probably.

Out of curiosity, I just remembered the existence of [JSON pointers](https://tools.ietf.org/html/rfc6901). Reliable syntax, but too much loss on easiness of use versus added complexity.

Never had any other use case and kept that function strictly in the application domain, with all its caveats clearly explained in docs.

@kurtmilam and @adispring: I'm curious about other use cases as well.

**Comment by kurtmilam** [comment]
To be clear, I'm not pushing for an inclusion of this function in the library. I'm fine with continuing to use a home-brewed version where necessary, and I think it's great that there's a recipe in the cookbook in case anyone else needs it.

In case you're still interested :D :

@CrossEye I'm not currently worried about the possibility of collisions with keys containing dots because my application generates the data it consumes, but I have given that some thought. It's one of the problems that I think makes a function like this of questionable value in a library like Ramda.

@CrossEye & @ascartabelli: I've seen this used twice. The first time was on an ASP.NET project I consulted on a decade ago. I believe they were using it just like @ascartabelli was. I helped them with the logic but never worked with it, myself.

More recently, I've used a version of the function that returns a key/value pair for every node in the input object in a pet project that does a lot with JSON configs. One 'benefit' I get out of it is the ability to separate the recursive walking of the object from the processing of the leaf node values, but that's probably just a cheap trick.

Where I really **need** it, or at least haven't figured out a different solution, is as a way to store information about every node in the input object. In this case, it's Promises. I build up a sort of Promise tree, where every node's Promise waits on the Promises of its children. That probably wouldn't be too useful on its own, but in my application, a node may also need to wait on Promises of other nodes that aren't its children.
```
        n1
      /    \
 waits on   \
    /        \
root     n2 waits on n1
    \        /
 waits on   /
      \    /
        n2
```

I just build up the Promise tree and fire `Promise.prop` (using bluebird), and everything is done in the correct order.

Lastly, I also use Handlebars.js in this project, and Handlebars.js refers to nodes in an object using the exact same dot-separated paths, so that's a bit of serendipity.

Hope I didn't bore anyone. I posted the diagram and details partly because I've been trying to figure out another way to do what I'm doing, and would be interested in any bright ideas that this exceedingly bright group might have :)

**Comment by ascartabelli** [comment]
> Hope I didn't bore anyone. I posted the diagram and details partly because I've been trying to figure out another way to do what I'm doing, and would be interested in any bright ideas that this exceedingly bright group might have :)

You haven't bored me, but it's hard for me to suggest something at a glance in this case.
Your promise tree looks very odd and it's not clear why, for example, n2 awaits for n1 and not vice versa. Nor is clear how results are consumed, if they are consumed at all: maybe it's only a chain of side effects?

So, random thoughts:

- can't you walk your input object once and build an array of Futures (rather than Promises: laziness may come in handy here) in the correct order and then run them as a sequence?
- not sure it fits your situation, but the saga pattern, or similar ones, are worth exploring to get some ideas: started using it in Redux with [redux-saga](https://github.com/yelouafi/redux-saga).
- can you boil it down to some form of dispatching? The [Flux dispatcher](https://github.com/facebook/flux/blob/master/src/Dispatcher.js) is an interesting alternative to the usual EventBus. Mainly because "callbacks can be deferred in whole or part until other callbacks have been executed".

**Comment by kurtmilam** [comment]
@ascartabelli thanks for the tips - I have some reading to do.

I don't want to go too far off reservation here, so if this isn't an appropriate conversation for this forum, I'll be happy to cut it off. It does apply to the question of how such a `flattenObj` function may be useful, however, so I'll continue a little unless someone indicates otherwise :)

In my config, it's possible to define, for each leaf node, dependencies on other nodes in the config. So in my simplistic example, `n1` would have a dependency on `n2` defined in the config. I use Handlebars.js to define the dependencies, so the initial value in `n1` might be something like `{{n2}}`.

Imagine `n2` is an http call to some REST API, and I want to extract a value from the return and put it in `n1` (`n1` may define another call to a REST API that needs to wait on a value returned by the `n2` call, for instance).

My application is basically a stab at using configurations to define and sequence multiple, related API calls. My odd (and I agree with you there) Promise tree construction takes care of firing every call as soon as the data needed to make that call has come in. Calls are made in parallel when appropriate, and sequentially when necessary. So you're right - the purpose of the config is to define multiple interdependent IO operations that are side effects.

**Comment by CrossEye** [comment]
> I don't want to go too far off reservation here, so if this isn't an appropriate conversation for this forum, I'll be happy to cut it off.

I for one am fascinated, although I have little to add.

I don't see any reason to move this from the Cookbook to the library, but I am glad that it's around for others with similar requirements.

**Comment by kurtmilam** [comment]
A new term has entered my vocabulary recently - "[Directed Acyclic Graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)" or DAG for short. What I called a "Promise Tree" earlier in this discussion is an example of a way to asynchronously process the individual tasks in a workflow represented by a DAG. Here's a library on npm that takes a different approach than I have taken to solve the same problem: https://github.com/jamiebuilds/task-graph-runner .
